### DISCUSSION TITLE
How to utilize service providers in Adonis v5?

### DISCUSSION THREAD

**Discussion by msrumon** [adonisjs/core#1421] [Help] [answered]
I need a service provider where I want to expose some methods to be used by other controllers. Right now I have stuffed all these logic inside controllers, making the controllers like 500-600 lines each and lots of duplicate codes all around the controllers. Due to lack of documentation, I am at a loss and unable to create providers. Say, I would like to create a `PaypalProvider` where I would like to have these methods: `createOrder()`, `getOrder()`, `captureOrder()` etc. Can someone help me please?

**Comment by thetutlage** [comment, marked as answer]
And why this can't live within a regular Javascript class? And then just import it inside your controller?

Yes, there aren't any docs on providers. But you can check the packages source code to get some sense of them until docs are written

**Reply by msrumon** [reply]
Do you want me to instantiate the objects all over the controllers? These providers should be instantiated once, and all controllers should use that single instance.

**Reply by thetutlage** [reply]
Writing the code as following will also give you a singleton

```ts
class Foo {
  public bar () {}
}

export default new Foo()
```

Inside controller (Assuming the `Foo` class is in `app/Services/Foo.ts` file)

```ts
import Foo from 'App/Services/Foo'

Foo.bar()
```

**Reply by msrumon** [reply]
```ts
class Demo {
  method() {
    return 'hello world';
  }
}

export default new Foo();
```
Which one is correct import in this case?
```ts
// (1) import Demo from '...';
// (2) import Foo from '...';
```

**Reply by thetutlage** [reply]
I believe, you need spend some time learning the basics of Javascript, ESM and Typescript. All this really has nothing to do with AdonisJS at all.

- How can you instantiate `Foo`, when the class name is `Demo`
- It doesn't matter what you name the import variable, when you have used `export default`.

**Reply by msrumon** [reply]
Thank you. Dumb question. :D

**Reply by msrumon** [reply]
In the case of https://github.com/adonisjs/core/discussions/1421#discussioncomment-48453, `import Env from '@ioc:Adonis/Core/Env'` throws `E_IOC_LOOKUP_FAILED` error. Any suggestion?
